src/components/node-list/node-list-items.js

```diff
@@ -330,3 +349,314 @@ export const getFilteredItems = createSelector(
+    getFilteredNodeModularPipelines,
+    (state) => state.modularPipelines,
+  ],
+  (
+    filteredModularPipelines,
+    filteredNodeModularPipelines,
+    modularPipelines
+  ) => {
+    const filteredMoudularPipelineParents = [];
+    const filteredModularPipeline = filteredModularPipelines.modularPipeline;
+
+    // 1. extract only modular pipelines with additional namespace
+    const childrenModularPipelines = filteredModularPipeline.filter(
+      (modularPipeline) => modularPipeline.id.includes('.')
+    );
+
+    // extract the parents only for those modular pipelines that does not have a filtered parent
+    childrenModularPipelines.forEach((childrenModularPipeline) => {
+      const levels = childrenModularPipeline.id.match(/\./g)
+        ? childrenModularPipeline.id.match(/\./g).length
+        : 0;
+
+      let lastIndex = 0;
+
+      for (let i = 0; i <= levels - 1; i++) {
+        // obtain the name of that pipeline
+        let parent = childrenModularPipeline.id.substr(
+          0,
+          childrenModularPipeline.id.indexOf('.', lastIndex)
+        );
+
+        // check against the filtered modular pipeline, existing list of parent pipelines and also the filtered node parent list
+        if (
+          !filteredModularPipeline.some(
+            (modularPipeline) => modularPipeline.id === parent
+          ) &&
+          !filteredMoudularPipelineParents.some(
+            (modularPipeline) => modularPipeline.id === parent
+          ) &&
+          !filteredNodeModularPipelines.some(
+            (modularPipeline) => modularPipeline.id === parent
+          )
+        ) {
+          // add the relevant modular pipeline to the list of parents
+          filteredMoudularPipelineParents.push(
+            // construct the item needed and then add it to the list
+            constructModularPipelineItem(
+              modularPipelines.filter(
+                (rawModularPipeline) => rawModularPipeline.id === parent
+              )[0]
+            )
+          );
+        }
+        lastIndex = childrenModularPipeline.id.indexOf('.', lastIndex) + 1;
+      }
+    });
+
+    return filteredMoudularPipelineParents;
+  }
+);
+
+/**
+ * returns the corresponding final set of modular pipelines
+ * for constructing the final nested tree list
+ */
+export const getFilteredTreeItems = createSelector(
+  [
+    getFilteredModularPipelineItems,
+    getFilteredNodeModularPipelines,
+    getFilteredModularPipelineParent,
+    (state) => state.modularPipelines,
+  ],
+  (
+    modularPipelineItems,
+    nodeModularPipelines,
+    modularPipelineParent,
+    modularPipelines
+  ) => {
+    modularPipelineItems = modularPipelineItems.modularPipeline;
+
+    let finalModularPipelines = [];
+
+    // sort all 3 sets of modular pipelines according to the original order
+    modularPipelines?.forEach((modularPipeline) => {
+      // check filtered modular pipelines
+      if (
+        modularPipelineItems.some(
```

Comment:
Similar to earlier, I think pull these conditions out of the `if` statements and give them a variable with a helpful name otherwise it's really hard to see what's a condition and what's assignment etc.